src/script_opt/ScriptOpt.cc

```diff
@@ -56,7 +56,7 @@ bool is_lambda(const ScriptFunc* f) { return lambdas.count(f) > 0; }
 
 bool is_when_lambda(const ScriptFunc* f) { return when_lambdas.count(f) > 0; }
 
-const FuncInfo* analyze_global_stmts(Stmt* stmts) {
+size_t analyze_global_stmts(Stmt* stmts) {
```

Comment:
First thought: Could this new contract be avoided by handing out a pointer to a persistent instance of FuncInfo for global stmts that is updated after optimization with the contents of whats in `funcs[global_stmts_handle]`?

Hmm, after trying, that's still weird, but same as before.. The other thought I'd have is let go of `analyze_global_stmts()` and fold it into `analyze_scripts()` via an extra optional stmt pointer for the global statements and let that return the optimized init_stmts entry point instead.

That should remove the need for a contract completely. The new function with a parameter where there's only one valid way to call it seems nice.

---
```
diff --git a/src/script_opt/ScriptOpt.cc b/src/script_opt/ScriptOpt.cc
index bb203277c..d17e2f885 100644
--- a/src/script_opt/ScriptOpt.cc
+++ b/src/script_opt/ScriptOpt.cc
@@ -37,6 +37,9 @@ static std::unordered_map<const ScriptFunc*, LambdaExpr*> lambdas;
 static std::unordered_set<const ScriptFunc*> when_lambdas;
 static ScriptFuncPtr global_stmts;
 
+static size_t global_stmts_handle = -1;
+static FuncInfo global_stmts_func_info = {nullptr, nullptr, nullptr, 0};
+
 void analyze_func(ScriptFuncPtr f) {
     // Even if we're analyzing only a subset of the scripts, we still
     // track all functions here because the inliner will need the full list.
@@ -56,7 +59,7 @@ bool is_lambda(const ScriptFunc* f) { return lambdas.count(f) > 0; }
 
 bool is_when_lambda(const ScriptFunc* f) { return when_lambdas.count(f) > 0; }
 
-size_t analyze_global_stmts(Stmt* stmts) {
+FuncInfo* analyze_global_stmts(Stmt* stmts) {
     // We ignore analysis_options.only_{files,funcs} - if they're in use, later
     // logic will keep this function from being compiled, but it's handy
     // now to enter it into "funcs" so we have a FuncInfo to return.
@@ -72,11 +75,11 @@ size_t analyze_global_stmts(Stmt* stmts) {
     global_stmts->AddBody(stmts->ThisPtr(), empty_inits, sc->Length());
 
     funcs.emplace_back(global_stmts, sc, stmts->ThisPtr(), 0);
-    return funcs.size() - 1;
+    global_stmts_handle = funcs.size() - 1;
+    global_stmts_func_info = funcs[global_stmts_handle];
+    return &global_stmts_func_info;
 }
 
-const FuncInfo& get_global_stmts(size_t global_ind) { return funcs[global_ind]; }
-
 void add_func_analysis_pattern(AnalyOpt& opts, const char* pat) {
     try {
         std::string full_pat = std::string("^(") + pat + ")$";
@@ -595,6 +598,9 @@ void analyze_scripts(bool no_unused_warnings) {
 
     if ( reporter->Errors() > 0 )
         reporter->FatalError("Optimized script execution aborted due to errors");
+
+    // May have been replaced/updated.
+    global_stmts_func_info = funcs[global_stmts_handle];
 }
```

Reply:
I though getting rid of the "conditional two-step API" might be nice, but this is all detail/zeek-setup.cc and works, so :man_shrugging:

Reply:
Thanks again for the help!

> getting rid of the "conditional two-step API" might be nice

Ah I see now. Yeah, would be slightly tidier to pass in the global statements to the call to `analyze_scripts()`. Will keep this in mind as a tweak next time I touch `zeek-setup.cc`. (The need for the special call later to execute the global statements remains, however, so it will still have a somewhat unobvious API.)

Reply:
> The need for the special call later to execute the global statements remains, however, so it will still have a somewhat unobvious API.

That's where I thought `analyzer_scripts()` could just return them (they might be the original ones, too). I might have missed another optimization related call in zeek-setup.cc though - you have a better picture there.